### PR TITLE
initial commit of conditions in all controllers

### PR DESCRIPTION
**Pull request by jmickey**
This PR adds `conditions` to the `ocm-controller`. Right now we're just setting the `Ready`, `Stalled`, and `Reconciling` conditions, but we can add additional support more conditions later.

Uses `flux.PatchHelper` to patch after the `Reconcile` function is returned. 

- [x] Add initial conditions `ready`, `reconciling`, `stalled`
- [x] Add test cases

### CODE REVIEW

File: api/v1alpha1/componentversion_types.go [hunk 1]
```diff
@@ -102,6 +102,23 @@ type ComponentVersionStatus struct {
...
+
+// GetConditions returns the conditions of the ComponentVersion.
+func (in *ComponentVersion) GetConditions() []metav1.Condition {
+	return in.Status.Conditions
+}
+
+// SetConditions sets the conditions of the ComponentVersion.
+func (in *ComponentVersion) SetConditions(conditions []metav1.Condition) {
+	in.Status.Conditions = conditions
 }
```
Comment (phoban01) [inline api/v1alpha1/componentversion_types.go:122]: Can we add a kubebuilder annotation here to show the Ready condition as a status field via kubectl?

Something like this:
```golang
// +kubebuilder:printcolumn:name="Ready",type="string",JSONPath=".status.conditions[?(@.type==\"Ready\")].status",description=""
// +kubebuilder:printcolumn:name="Status",type="string",JSONPath=".status.conditions[?(@.type==\"Ready\")].message",description=""

File: controllers/componentversion_controller.go [hunk 2]
```diff
@@ -154,35 +185,78 @@ func (r *ComponentVersionReconciler) checkVersion(ctx context.Context, obj *v1al
...
-func (r *ComponentVersionReconciler) reconcile(ctx context.Context, obj *v1alpha1.ComponentVersion, version string) (ctrl.Result, error) {
+func (r *ComponentVersionReconciler) reconcile(ctx context.Context, obj *v1alpha1.ComponentVersion, version string) (result ctrl.Result, retErr error) {
 	log := log.FromContext(ctx).WithName("ocm-component-version-reconcile")
 
-	// get component version
-	// TODO@souleb: resolve all referenced dependencies and pass them to the function
-	// this will allow us to return early if the dependencies are not ready, and set error
-	// conditions on the component version.
-	// We can the pass the needed dependencies to the function as Option
+	defer func() {
```
Comment (phoban01) [inline controllers/componentversion_controller.go:191]: is there any need for this defer in the inner reconcile method? Could it not be handled in the outer reconcile?

**Review by phoban01 — COMMENTED** [review]
Just a couple of comments, overall looks great.

**Comment by Skarlso** [comment]
Created component version, seems to be reconciling fine after setting requeue to 10s.

```yaml
Status:
  Component Descriptor:
    Component Descriptor Ref:
      Name:       github.com-acme-podinfo-v2.0.0-9496839999115680383
      Namespace:  ocm-system
    Name:         github.com/acme/podinfo
    Version:      v2.0.0
  Conditions:
    Last Transition Time:  2023-02-07T13:00:53Z
    Message:               Reconciliation success
    Observed Generation:   2
    Reason:                Succeeded
    Status:                True
    Type:                  Ready
  Observed Generation:     2
  Reconciled Version:      v2.0.0
Events:                    <none>

```

Testing now if I update the component version and it reconciles a new version, does the generation increase.

**Comment by Skarlso** [comment]
Snapshot looks okay.
```yaml
Status:
  Conditions:
    Last Transition Time:  2023-02-07T13:21:50Z
    Message:               Snapshot with name 'podinfo-deployment' is ready
    Observed Generation:   1
    Reason:                Succeeded
    Status:                True
    Type:                  Ready
  Digest:                  sha256:d306be7092d2bbdb329de698b60dbb4728db346a1d2f77a094e7bf795a7f5acf
  Observed Generation:     1
  Repository URL:          http://registry.ocm-system.svc.cluster.local:5000/sha-6329374125460483158
  Tag:                     latest
Events:                    <none>

```

Resource also has been set:

```yaml
Status:
  Conditions:
    Last Transition Time:  2023-02-07T13:21:50Z
    Message:               Reconciliation success
    Observed Generation:   1
    Reason:                Succeeded
    Status:                True
    Type:                  Ready
  Observed Generation:     1
Events:                    <none>
```

**Comment by Skarlso** [comment]
Failed snapshot doesn't have a status yet as that code part will be added in a followup PR.

**Review by phoban01 — APPROVED** [review]

**Pull request merged**